src/controller/CHIPDeviceController.cpp

```diff
@@ -227,47 +231,90 @@ CHIP_ERROR DeviceController::Init(NodeId localDeviceId, ControllerInitParams par
+CHIP_ERROR DeviceController::GenerateOperationalCertificates(const ByteSpan & CSR, NodeId deviceId, MutableByteSpan & cert)
 {
-    ChipLogProgress(Controller, "Getting operational keys");
-    Crypto::P256Keypair * keypair = admin->GetOperationalKey();
+    // This code requires about 2K RAM to generate the certificates.
+    // The code would run as part of commissioner applications, so RAM requirements should be fine.
+    // Need to analyze if this requirement could be better managed by using static memory pools.
+    chip::Platform::ScopedMemoryBuffer<uint8_t> noc;
+    ReturnErrorCodeIf(!noc.Alloc(kMaxCHIPDERCertLength), CHIP_ERROR_NO_MEMORY);
 
-    ReturnErrorCodeIf(keypair == nullptr, CHIP_ERROR_NO_MEMORY);
+    uint32_t nocLen = 0;
+    ChipLogProgress(Controller, "Generating operational certificate for device " ChipLogFormatX64, ChipLogValueX64(deviceId));
+    ReturnErrorOnFailure(mOperationalCredentialsDelegate->GenerateNodeOperationalCertificate(
+        PeerId().SetNodeId(deviceId), CSR, 1, noc.Get(), kMaxCHIPDERCertLength, nocLen));
 
-    if (!admin->AreCredentialsAvailable())
+    ReturnErrorCodeIf(nocLen == 0, CHIP_ERROR_CERT_NOT_FOUND);
+
+    chip::Platform::ScopedMemoryBuffer<uint8_t> ica;
+    ReturnErrorCodeIf(!ica.Alloc(kMaxCHIPDERCertLength), CHIP_ERROR_NO_MEMORY);
+
+    uint32_t icaLen = 0;
+    ChipLogProgress(Controller, "Getting intermediate CA certificate from the issuer");
+    CHIP_ERROR err = mOperationalCredentialsDelegate->GetIntermediateCACertificate(0, ica.Get(), kMaxCHIPDERCertLength, icaLen);
+    ChipLogProgress(Controller, "GetIntermediateCACertificate returned %" PRId32, err);
+    if (err == CHIP_ERROR_INTERMEDIATE_CA_NOT_REQUIRED)
     {
-        chip::Platform::ScopedMemoryBuffer<uint8_t> buffer1;
-        ReturnErrorCodeIf(!buffer1.Alloc(kMaxCHIPCSRLength), CHIP_ERROR_NO_MEMORY);
+        // This implies that the commissioner application uses root CA to sign the operational
+        // certificates, and an intermediate CA is not needed. It's not an error condition, so
+        // let's just send operational certificate and root CA certificate to the device.
+        icaLen = 0;
+        ChipLogProgress(Controller, "Intermediate CA is not needed");
+    }
+    else if (err != CHIP_NO_ERROR)
+    {
+        return err;
+    }
 
-        chip::Platform::ScopedMemoryBuffer<uint8_t> buffer2;
-        ReturnErrorCodeIf(!buffer2.Alloc(kMaxCHIPOpCertLength), CHIP_ERROR_NO_MEMORY);
+    ReturnErrorOnFailure(ConvertX509CertsToChipCertArray(ByteSpan(noc.Get(), nocLen), ByteSpan(ica.Get(), icaLen), cert));
 
-        uint8_t * CSR    = buffer1.Get();
-        size_t csrLength = kMaxCHIPCSRLength;
-        ReturnErrorOnFailure(keypair->NewCertificateSigningRequest(CSR, csrLength));
+    return CHIP_NO_ERROR;
+}
 
-        uint8_t * cert   = buffer2.Get();
-        uint32_t certLen = 0;
+CHIP_ERROR DeviceController::LoadLocalCredentials(Transport::AdminPairingInfo * admin)
```

Comment:
Currently, we have a design that is strictly single-fabric on the controller side.  I would argue that this is not consistent with the spec.  Multi-admin support should be symmetric between controllers and accessories.

We could either address this by allowing instantiation of multiple controller objects (definitely not supported today), or we could make the controller multi-fabric capable.  The latter approach would require things like e.g. passing `AdminPairingTable` here instead of `AdminPairingInfo`, and reworking the device pool to index on (root CA public key, fabric ID, node ID) or (fabric Reference, node ID) instead of just node ID.

I know it's out of scope for this PR.  But I'm raising the issue.  I don't see how we can get to a spec-compliant 1.0 without such support.